### PR TITLE
*: Add AWS lambda to Backend API

### PR DESCRIPTION
**Pull request by timflannagan**
# Description

Overhaul the AWS lambda fields in the Backend API. Implements the API-related
changes outlined in #10644.

Related to #10665.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: api/v1alpha1/backend_types.go [hunk 1]
```diff
@@ -61,29 +61,91 @@ type BackendSpec struct {
...
-	SecretRef corev1.LocalObjectReference `json:"secretRef,omitempty"`
+	Region *string `json:"region,omitempty"`
 }
 
-// StaticBackend is the static backend configuration.
-type StaticBackend struct {
-	// Hosts is the list of hosts.
+// AwsAuth defines the authentication method to use for the backend.
+// +kubebuilder:validation:XValidation:message="only one auth method can be specified",rule="!(has(self.irsa) && has(self.secret))"
+type AwsAuth struct {
```
Comment (timflannagan) [inline api/v1alpha1/backend_types.go:82]: I previously had a dedicated `type AwsAuthType string` field, but removed that to make the spec.auth field an optional field to configure. I'm not sure whether this struct's current form is now a bit awkward now:

- You either omit spec.aws.auth and inherit the default (use the credential flow outlined in https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/aws_lambda_filter#credentials), or
- Configure spec.auth.aws.[irsa, secret] for fine-grain control over how a lambda backend is authenticated

One potential gap is I primarily tested the default method on an EKS cluster and users' may want to route to lambda backends in different cluster topologies. Additionally, we may want to expose the "default" behavior as a field on the AwsAuth struct for uniformity, but 1.) I don't want this behavior to be codified as a bool (e.g. `useDefaultCredentialsProvider`, and 2.) it might be difficult to expose the right knobs with lambda being the primary AWS service use case.

File: api/v1alpha1/backend_types.go [hunk 2]
```diff
@@ -61,29 +61,91 @@ type BackendSpec struct {
...
+	// IRSA is the IRSA authentication configuration.
+	// +optional
+	IRSA *AWSAuthIRSA `json:"irsa,omitempty"`
+	// Secret is a reference to a secret containing AWS credentials.
+	// +optional
+	Secret *corev1.LocalObjectReference `json:"secret,omitempty"`
```
Comment (timflannagan) [inline api/v1alpha1/backend_types.go:88]: Used the corev1.LocalObjectReference type to be consistent with the BackendAI type. Previously a wrapper type was used to decouple the API from an external dependency.

File: api/v1alpha1/backend_types.go [hunk 3]
```diff
@@ -61,29 +61,91 @@ type BackendSpec struct {
...
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=2048
+	EndpointURL string `json:"endpointURL,omitempty"`
+	// FunctionName is the name of the lambda function to invoke.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=140
+	FunctionName string `json:"functionName"`
+	// InvocationMode defines how to invoke the lambda function.
+	// Defaults to SYNCHRONOUS if not specified.
```
Comment (timflannagan) [inline api/v1alpha1/backend_types.go:119]: FIXME.

File: internal/kgateway/translator/gateway/testutils/inputs/http-with-lambda-destination/manifests.yaml
Comment (timflannagan) [inline internal/kgateway/translator/gateway/testutils/inputs/http-with-lambda-destination/manifests.yaml:1]: FIXME: failing unit test.

File: api/v1alpha1/backend_types.go [hunk 4]
```diff
@@ -61,29 +61,91 @@ type BackendSpec struct {
...
+	Qualifier string `json:"qualifier,omitempty"`
+}
+
+// StaticBackend is an upstream that references a static list of hosts.
+type StaticBackend struct {
+	// Hosts is a list of hosts to use for the upstream.
+	// +kubebuilder:validation:required
 	// +kubebuilder:validation:MinItems=1
+	// +listType=map
+	// +listMapKey=host
```
Comment (yuval-k) [inline api/v1alpha1/backend_types.go:137]: not sure if map with key host is true; you could in theory have the same host and two ports as valid config

File: api/v1alpha1/backend_types.go [hunk 5]
```diff
@@ -61,29 +61,91 @@ type BackendSpec struct {
...
+	// +optional
+	Auth *AwsAuth `json:"auth,omitempty"`
+	// Lambda configures the AWS lambda service.
 	// +optional
-	Region string `json:"region,omitempty"`
-	// SecretRef is the secret reference for the AWS credentials.
+	Lambda *AwsLambda `json:"lambda,omitempty"`
+	// Region is the AWS region to use for the upstream.
+	// Defaults to us-east-1 if not specified.
 	// +optional
```
Comment (yuval-k) [inline api/v1alpha1/backend_types.go:76]: consider adding `+kubebuilder:default=us-east-1` ?

File: api/v1alpha1/backend_types.go (hunk 1 above)
Comment (yuval-k) [reply api/v1alpha1/backend_types.go:82]: `AwsAuthType` can be `Default|Irsa|Secret` defaulting to Default?

File: api/v1alpha1/backend_types.go [hunk 6]
```diff
@@ -61,29 +61,125 @@ type BackendSpec struct {
...
+// AwsAuth defines the authentication method to use for the backend.
+// +union
+// +kubebuilder:validation:XValidation:message="irsa auth must be nil if the type is not 'irsa'",rule="!(has(self.irsa) && self.type != 'irsa')"
+// +kubebuilder:validation:XValidation:message="irsa auth must be specified when type is 'irsa'",rule="!(!has(self.irsa) && self.type == 'irsa')"
+// +kubebuilder:validation:XValidation:message="secret auth must be nil if the type is not 'secret'",rule="!(has(self.secret) && self.type != 'secret')"
+// +kubebuilder:validation:XValidation:message="secret auth must be specified when type is 'secret'",rule="!(!has(self.secret) && self.type == 'secret')"
+type AwsAuth struct {
+	// Type is the type of authentication to use for the backend.
+	// +optional
+	// +kubebuilder:validation:Optional
```
Comment (timflannagan) [inline api/v1alpha1/backend_types.go:110]: Included both of these markers since we use other codegen tooling: https://book.kubebuilder.io/reference/markers. We might be able to standardize on one of the optional markers.

File: api/v1alpha1/backend_types.go [hunk 7]
```diff
@@ -61,29 +61,125 @@ type BackendSpec struct {
...
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=12
+	AccountId string `json:"accountId"`
+	// Auth specifies the authentication method to use for the backend.
+	// When unspecified, the default authentication method will be used.
 	// +optional
-	Region string `json:"region,omitempty"`
-	// SecretRef is the secret reference for the AWS credentials.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default={type: default}
```
Comment (timflannagan) [inline api/v1alpha1/backend_types.go:73]: Needed to provide the "default" enum type on the AwsAuth type to prevent panics in the code.

File: api/v1alpha1/backend_types.go [hunk 8]
```diff
@@ -61,29 +61,125 @@ type BackendSpec struct {
 
 // AwsBackend is the AWS backend configuration.
 type AwsBackend struct {
-	// Region is the AWS region.
+	// AccountId is the AWS account ID to use for the backend.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=12
```
[the remark this answers is not in the document]
Comment (yuval-k) [reply api/v1alpha1/backend_types.go:67]: do we want to add a regex/cel to enforce that it is digits?

[the remark this answers is not in the document]
Comment (timflannagan) [reply api/v1alpha1/backend_types.go:67]: Good idea. Had the staged locally on my feature branch. I'll cherry-pick it on this branch.

[the remark this answers is not in the document]
Comment (timflannagan) [reply api/v1alpha1/backend_types.go:67]: Just pushed up some stricter validation including the accoundID field:

```bash
$ k apply -f test/kubernetes/e2e/features/lambda/testdata/lambda-backend.yaml
httproute.gateway.networking.k8s.io/lambda-route configured
The Backend "lambda-backend" is invalid: spec.aws.accountId: Invalid value: "A00000000000": spec.aws.accountId in body should match '^[0-9]{12}$'
```

File: api/v1alpha1/backend_types.go [hunk 9]
```diff
@@ -61,29 +61,126 @@ type BackendSpec struct {
...
+)
+
+// AwsLambda configures the AWS lambda service.
+type AwsLambda struct {
+	// EndpointURL is the URL or domain for the Lambda service. This is primarily
+	// useful for testing and development purposes. When omitted, the default
+	// lambda hostname will be used.
 	// +optional
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Pattern="^https?://[-a-zA-Z0-9@:%.+~#?&/=]+$"
```
Comment (timflannagan) [inline api/v1alpha1/backend_types.go:148]: Add validation marker to help prevent malformed URLs from being accepted by the apiserver.

File: api/v1alpha1/backend_types.go [hunk 10]
```diff
@@ -61,29 +61,128 @@ type BackendSpec struct {
...
+	// +kubebuilder:validation:MaxLength=2048
+	EndpointURL string `json:"endpointURL,omitempty"`
+	// FunctionName is the name of the Lambda function to invoke.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern="^[A-Za-z0-9-_]{1,140}$"
+	FunctionName string `json:"functionName"`
+	// InvocationMode defines how to invoke the Lambda function.
+	// Defaults to Sync.
```
Comment (timflannagan) [inline api/v1alpha1/backend_types.go:156]: Is this the right default?

File: api/v1alpha1/backend_types.go (hunk 6 above)
Comment (lgadban) [reply api/v1alpha1/backend_types.go:115]: I'm slightly confused on this being optional.
if you don't specify an AwsAuth at all it will use the default,
but it will also default if you provide e.g. an empty AwsAuth?

File: api/v1alpha1/backend_types.go [hunk 11]
```diff
@@ -61,29 +61,128 @@ type BackendSpec struct {
...
+// StaticBackend references a static list of hosts.
+type StaticBackend struct {
+	// Hosts is a list of hosts to use for the backend.
+	// +kubebuilder:validation:required
 	// +kubebuilder:validation:MinItems=1
 	Hosts []Host `json:"hosts,omitempty"`
 }
 
-// Host is a host and port pair.
+// Host defines a static upstream host.
```
Comment (lgadban) [inline api/v1alpha1/backend_types.go:182]:
```suggestion
// Host defines a static backend host.
```

File: api/v1alpha1/backend_types.go [hunk 12]
```diff
@@ -61,29 +61,128 @@ type BackendSpec struct {
...
 type Host struct {
-	// Host is the host name.
+	// Host is the host to use for the backend.
```
Comment (lgadban) [inline api/v1alpha1/backend_types.go:184]: why remove host name? because it can be an IP addr?

File: internal/kgateway/translator/gateway/testutils/inputs/http-with-lambda-destination/manifests.yaml
Comment (lgadban) [inline internal/kgateway/translator/gateway/testutils/inputs/http-with-lambda-destination/manifests.yaml:1]: anything currently using this file?

Comment (timflannagan) [reply internal/kgateway/translator/gateway/testutils/inputs/http-with-lambda-destination/manifests.yaml:1]: yeah there's an existing gateway translator test case that reads this input file. might be cruft if we'll have envtest coverage for several lambda Backend edge cases coming as a follow-up implementation.

File: api/v1alpha1/backend_types.go (hunk 6 above)
Comment (timflannagan) [reply api/v1alpha1/backend_types.go:119]: right we're relying on CRD defaulting https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting to add those fields.

Comment (timflannagan) [reply api/v1alpha1/backend_types.go:119]: if the UX is confusing, then maybe we should reconsider making auth a required field. here's a couple of inline thoughts:

- backwards compatible if we want to make auth optional in the future
- another field that uses need to configure, which has UX implications
- potentially more "declarative" or explicit, which could lead to less complexity in the code

File: api/v1alpha1/backend_types.go [hunk 13]
```diff
@@ -61,29 +61,133 @@ type BackendSpec struct {
...
+	// AccountId is the AWS account ID to use for the backend.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=12
+	// +kubebuilder:validation:Pattern="^[0-9]{12}$"
+	AccountId string `json:"accountId"`
+	// Auth specifies an explicit AWS authentication method for the backend.
+	// When omitted, the authentication method will be inferred from the
+	// environment (e.g. instance metadata, EKS Pod Identity, environment variables, etc.)
+	// This may not work in all environments, so it is recommended to specify an authentication method.
```
Comment (lgadban) [inline api/v1alpha1/backend_types.go:73]: I'm not sure this is accurate, pod identity could be perfectly valid right?
Not a big deal though

File: api/v1alpha1/backend_types.go (hunk 6 above)
Comment (lgadban) [reply api/v1alpha1/backend_types.go:121]: This is exactly what I was envisioning. I like it!

**Review by lgadban — APPROVED** [review]
